File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/PropertyBinding.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.EventArgs;
import com.hello.uxml.tools.framework.events.EventHandler;
import com.hello.uxml.tools.framework.events.EventNotifier;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Defines a binding of a target object's property to a chain of properties
 * starting from a source object.
 *
 * @author sanjayc@ (Sanjay Chouksey)
 * @author ferhat
 */
public class PropertyBinding {
  private Object targetObject;
  private Object targetProperty;
  private List<Object> sourceChain = new ArrayList<Object>();
  private Object[] propChain;
  private EventHandler listener;
  private ValueTransform valueTransform;
  private String transformArg;
  private static final String ERROR_MSG_NO_MATCHING_TARGET =
      "onChange: can't find source matching the event target.";
  private static final Logger logger = Logger.getLogger(PropertyBinding.class.getName());
  public static ValueTransform negateBoolean = new NegateBooleanTransform();

  /**
   * Constructor.
   */
  public PropertyBinding(UxmlElement targetObject, Object targetProperty,
      Object sourceObject, Object[] propertyChain) {
    this(targetObject, targetProperty, sourceObject, propertyChain, null);
  }

  /**
   * Constructor.
   */
  public PropertyBinding(UxmlElement targetObject, Object targetProperty,
      Object sourceObject, Object[] propertyChain, ValueTransform transform) {
    this(targetObject, targetProperty, sourceObject, propertyChain, transform, null);
  }

  /**
   * Constructor.
   */
  public PropertyBinding(UxmlElement targetObject, Object targetProperty,
      Object sourceObject, Object[] propertyChain, ValueTransform transform, String transformArg) {
    this.targetObject = targetObject;
    this.targetProperty = targetProperty;
    sourceChain.add(sourceObject);
    propChain = new Object[propertyChain.length];
    for (int i = 0; i < propertyChain.length; i++) {
      propChain[i] = propertyChain[i];
    }
    valueTransform = transform;
    this.transformArg = transformArg;
    bind(sourceObject, 0);
  }

  /**
   * Returns target object.
   */
  public Object getTargetObject() {
    return targetObject;
  }

  /**
   * Recursively binds to the source and listens for property change
   * event on it.
   *
   * @param source The object to listen for property change event.
   * @param propChainIndex Chain of properties starting from source.
   */
  private void bind(Object source, int propChainIndex) {

    // now read the first property in the chain
    Object propName = propChain[propChainIndex];

    // add event listener for property change and collection change events
    if (source instanceof EventNotifier) {
      EventNotifier notifier = (EventNotifier) source;
      PropertyDefinition propDef = null;
      // use propdefs if source is Element
      if (propName instanceof PropertyDefinition) {
        propDef = (PropertyDefinition) propName;
      } else {
        propDef = PropertyCache.getPropertyDefinition(
              ((UxmlElement) source).getClass(), (String) propName);
        if (propDef != null) {
          propChain[propChainIndex] = propDef;
        }
        propName = propDef;
      }

      // listen for change event on propName
      listener = new EventHandler() {
        @Override
        public void handleEvent(EventNotifier target, EventArgs e) {
          onPropChange((PropertyChangedEvent) e);
        }
      };
      notifier.addListener(propName, listener);

      //  TODO(ferhat) if source is model, also listen for collection change events
      //  if (source is Model) {
      //    notifier.addListener(CollectionChangedEvent.eventDef, onCollChange);
      //  }
    }

    // now get the property value
    Object prop = null;
    prop = getProperty(source, propName);
    if (prop == null) {
      setProperty(prop);
      return;
    }

    // if there are more items in prop chain, keep going
    if (propChainIndex < propChain.length - 1) {
      // store this source object in sourceChain
      sourceChain.add(prop);
      bind(prop, propChainIndex + 1);
    } else {
      // set the value of the target
      setProperty(prop);
    }
  }

  /**
   * Clear all bindings and remove all property change event listeners.
   */
  public void clear() {
    targetObject = null;
    targetProperty = null;
    sourceChain = null;
    propChain = null;
    unbind(0);
  }

  /**
   * Removes the binding for all source objects starting from specified index.
   * This removes the event listener for property change and also the object
   * from the source chain.
   */
  private void unbind(int index) {
    // remove event listeners from sources starting from this index
    for (int i = index; i < sourceChain.size(); i++) {
      Object source = sourceChain.get(i);
      String propName = (String) propChain[i];

      if (!(source instanceof EventNotifier)) {
        continue;
      }

      EventNotifier notifier = (EventNotifier) source;

      notifier.removeListener(propName, listener);
      listener = null;

      // TODO(ferhat) implement unbind for Model
      // if (source is Model) {
      //   notifier.removeListener(CollectionChangedEvent.eventDef, onCollChange);
      // }
    }
    // remove all objects starting from this source
    sourceChain.remove(index);
  }

  /**
   * Handles property change events by rebinding all source objects under the
   * object firing the event.
   */
  private void onPropChange(PropertyChangedEvent event) {
    // find the source object in sourceChain that fired this event
    int index = sourceChain.indexOf(event.getSource());
    if (index < 0 || index >= sourceChain.size()) {
      // this should never happen
      logger.log(Level.WARNING, ERROR_MSG_NO_MATCHING_TARGET);
      return;
    }

    // check if this property exists in our property chain
    if (!propChain[index].equals(event.getProperty())) {
        return; // this should never happen
    }

    // if this is the last prop in the chain, just update its value
    if (index == (sourceChain.size() - 1)) {
      setProperty(event.getNewValue());
    } else {
      // all objects underneath this source are dirty, unbind them
      unbind(index);

      // now re-bind to all new objects starting from this source
      bind(event.getSource(), index);
    }
  }

  // TODO(ferhat): implement Model rebinding
  //  /**
  //   * Handles collection change events by rebinding all source objects under
  //   * the object firing the event.
  //   */
  //  public void onCollChange(event:CollectionChangedEvent):void {
  //    // find the source object in sourceChain that fired this event
  //    var sourceIndex:int = sourceChain.indexOf(event.source);
  //  if (sourceIndex < 0 || sourceIndex >= sourceChain.length) {
  //    // this should never happen
  //    trace("WARN", "onChange: source not found matching the event target.");
  //    return;
  //  }
  //
  //  // check if this property exists in our property chain
  //  if (propChain[sourceIndex] != event.index.toString())
  //    return; // this should never happen
  //
  //  // all objects underneath this source are dirty, unbind them
  //  unbind(sourceIndex);
  //
  //  // now re-bind to all new objects starting from this source
  //  bind(event.source, sourceIndex);
  //  }

  /**
   * Gets the property on a source object.
   */
  private Object getProperty(Object source, Object propName) {
    // if source is Element, use propDefs to access the property
    if (source instanceof UxmlElement) {
      // check if propName is PropertyDefinition
      PropertyDefinition propDef = (PropertyDefinition) propName;

      if (propDef != null) {
        return ((UxmlElement) source).getProperty(propDef);
      } else {
      return null;
    }
    // TODO(ferhat):Model
    //} else if (source is Model) {
    //  return Model(source).getProperty(propName as String);
    } else {
      return Application.getCurrent().readDynamicValue(source, (String) propName);
    }
  }

  /** Sets the property on the target object to the specified value. */
  private void setProperty(Object value) {
    if (valueTransform != null) {
      value = valueTransform.transformValue(value, transformArg);
    }
    // if target object is Element, use propDefs to set the property
    if (targetObject instanceof UxmlElement) {

      // if targetPropDef is not cached, do so now
      if (targetProperty instanceof String) {
        targetProperty = PropertyCache.getPropertyDefinition(
            ((UxmlElement) targetObject).getClass(), (String) targetProperty);
      }
      if (targetProperty != null) {
        ((UxmlElement) targetObject).setProperty((PropertyDefinition) targetProperty, value);
      }
      // TODO(ferhat): Add model support for PropertyBinding.setProperty
      //    } else if (targetObject is Model) {
      //
      //      // if targetProp is numeric, it is an index otherwise a property
      //      var index:int = parseInt(targetProp as String);
      //    if (isNaN(index))
      //      Model(targetObject).setProperty(targetProp as String, value);
      //    else
      //      Model(targetObject).setItemAt(index, value);
    } else {
      Application.getCurrent().writeDynamicValue(targetObject, (String) targetProperty, value);
    }
  }

  /**
   * Implements negate boolean value transform for bindings.
   */
  public static class NegateBooleanTransform implements ValueTransform {
    @Override
    public Object transformValue(Object value, Object transArg) {
      if (value instanceof Boolean) {
        return !((Boolean) value);
      }
      return !Boolean.valueOf(String.valueOf(value));
    }
  }
}

File: framework/lib/property_binding.dart
part of uxml;

/**
 * Defines a binding of a target object's property to a chain of properties
 * starting from a source object.
 *
 * @author sanjayc@ (Sanjay Chouksey)
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class PropertyBinding {

  IKeyValue _targetObject;          // target object for this binding.
  IKeyValue _source;                // root source object.
  Object _targetProp;               // target property.
  List<Object> _propChain;          // property chain for this binding.
  List<Object> _sources;            // source objects we are listening to.
  BindingTransform _valueTransform; // transforms value during binding.
  String _transformArg;             // optional arg to transform function.
  Map<Object, EventHandler> sourceListeners;

  /**
   * Constructs a binding of source object to target object. For example, the
   * following code fragment:
   *    <code>textbox.text = album.metainfo.sizeinfo.duration;</code>
   * or markup:
   *    <textbox text="{album.metainfo.sizeinfo.duration}"/>
   *
   * results in:
   *    targetObject = textbox
   *    targetProp = textPropDef
   *    source = album
   *    propChain = ["metainfo", "sizeinfo", "duration"]
   *
   * The following types of source objects are supported:
   *    1. Classes derived from <code>UxmlElement</code>
   *    2. Objects of type <code>Model</code>
   *    3. Readonly mode: IKeyValue, read/sync: IElementEvents
   *
   * @param target The object to which this binding applies.
   * @param targetProp The property definition or string name of property.
   *                   This can be a dotted property name.
   * @param source The object to which the target is bound to.
   * @param propChain The chain of properties originating from the source.
   */
  PropertyBinding(IKeyValue target, Object targetProperty, IKeyValue source,
      List<Object> propertyChain, [BindingTransform valueTransform = null,
      String transformArg = null]) {
    _targetObject = target;
    _targetProp = targetProperty;
    if ((_targetProp is String) && (_targetObject is UxmlElement)) {
      String targetPath = _targetProp;
      if (targetPath.indexOf('.') == -1) {
        UxmlElement elm = _targetObject;
        PropertyDefinition propDef = elm.getDefinition().findProperty(
            _targetProp);
        if (propDef != null) {
          _targetProp = propDef;
        } else {
          String propName = _targetProp;
          Application.current.log("Find property failed on elementDef "
              "${elm.getDefinition().name} searching for $propName");
        }
      }
    }
    _source = source;
    _propChain = propertyChain;
    _valueTransform = valueTransform;
    _transformArg = transformArg;
    _sources = <IKeyValue>[];
    sourceListeners = new Map<Object, EventHandler>();

    String propName;

    // If source property chain is prefixed with parent. , bind
    // to parent element.
    if (propertyChain.length > 0 && propertyChain[0] is String) {
      propName = propertyChain[0];
      if (propName == "parent" && (source is UxmlElement)) {
        UxmlElement sourceElement = source;
        source = sourceElement.parent;
        propertyChain.removeAt(0);
      }
    }

    if (target == source) {
      if (propertyChain.length > 0 &&
          (propertyChain[0] is PropertyDefinition)) {
        PropertyDefinition targetProp = propertyChain[0];
        if ((targetProp != null) && ((targetProp.flags &
            PropertyFlags.INHERIT) != 0) &&
            targetProp == _targetProp) {
          UxmlElement sourceElement = source;
          source = sourceElement.parent;
          _source = source;
        }
      }
    }

    // initialize the source chain with root source.
    _sources.add(_source);

    // recursively walk the source chain to listen for property change
    // events and compute the initial value
    _bind(source, 0);
  }

  /**
   * Clear all bindings and remove all property change event listeners.
   */
  void clear() {
    _unbind(0);
    _targetObject = null;
    _targetProp = null;
    _source = null;
    _propChain = null;
  }

  /**
   * Recursively binds to the source and listens for property change
   * event on it.
   *
   * @param source The object to listen for property change event.
   * @param propChainIndex Index into chain of properties starting
   *   from source.
   */
  void _bind(Object source, int propChainIndex) {
    if (_propChain.length == 0) {
      _setTargetProperty(source);
      return;
    }
    // now read the first property in the chain
    Object propKey = _propChain[propChainIndex];

    // convert propName to source's property definition.
    if ((propKey is String) && (source is UxmlElement)) {
      UxmlElement elem = source;
      PropertyDefinition property = elem.getDefinition().findProperty(propKey);
      if (property == null) {
        print("Could not locate property $propKey on element type $elem");
      }
      _propChain[propChainIndex] = property;
      propKey = property;
    }

    // add event listener for property change and collection change events
    if (source is IElementEvents) {
      IElementEvents elementEvents = source;
      EventHandler closureVal = _onPropChange;
      if (sourceListeners[propChainIndex] != null) {
        elementEvents.removeListener(propKey, sourceListeners[propChainIndex]);
      }
      sourceListeners[propChainIndex] = closureVal;
      elementEvents.addListener(propKey, closureVal);
    }

    // now get the property value
    Object propValue;
    IKeyValue kv = source;
    try {
      propValue = kv.getProperty(propKey);

      if ((propKey is PropertyDefinition) && kv is UxmlElement) {
        UxmlElement src = kv;
        PropertyDefinition srcProp = propKey;
      }
    } on Exception catch (e) {
      propValue = null;
    }
    if (propValue == null) {
      // TODO(ferhat) this code makes sure that source does have a property by
      // propName, but it is needed for two-way binding to avoid turning both
      // source and target to null or undefined.
      if (kv != null && kv.overridesProperty(propKey) &&
          propChainIndex == (_propChain.length - 1)) {
        _setTargetProperty(propValue);
      }
      return;
    }

    // if there are more items in prop chain, keep going
    if (propChainIndex < (_propChain.length - 1)) {
      // store this source object in sourceChain
      _sources.add(propValue);
      _bind(propValue, propChainIndex + 1);
    } else {
      if (!(identical(propValue, PropertyDefaults.NO_DEFAULT))) {
        _setTargetProperty(propValue);
      }
    }
  }

  /**
   * Removes the binding for all source objects starting from specified index.
   * This removes the event listener for property change and also the object
   * from the source chain.
   */
  void _unbind(int index) {
    if (index >= _sources.length) {
      return;
    }
    // remove event listeners from sources starting from this index
    for (int i = index; i < _sources.length; i++) {
      Object valueSource = _sources[i];
      Object propKey = _propChain[i];

      if (valueSource is IElementEvents) {
        IElementEvents elementEvents = valueSource;
        elementEvents.removeListener(propKey, sourceListeners[i]);
        sourceListeners[i] = null;
      }
    }
    // remove all objects starting from this source
    _sources.removeRange(index, _sources.length);
  }

  /**
   * Handles property change events by rebinding all source objects under the
   * object firing the event.
   */
  void _onPropChange(EventArgs event) {
    if (!(event is PropertyChangedEvent)) {
      return;
    }
    PropertyChangedEvent propChangeEvent = event;

    // find the source object in sourceChain that fired this event
    int index = _sources.indexOf(event.source);
    if (index < 0 || index >= _sources.length) {
      // this should never happen
      List<IKeyValue> s = _sources;
      List<Object> pchain = _propChain;
      Application.current.log("warning: onChange: source not found matching "
          "the event target. ${event.source}");
      return;
    }

    // check if the source has changed (event.property == null) or
    // check if this property exists in our property chain
    if (propChangeEvent.property != null &&
        _propChain.indexOf(propChangeEvent.property, 0) < 0) {
      return;
    } // this should never happen

    // if this is the last prop in the chain, just update its value
    if (index == (_propChain.length - 1)) {
      _setTargetProperty(propChangeEvent.newValue);
    } else {
      // all objects underneath this source are dirty, unbind them
      _unbind(index + 1);
      // now re-bind to all new objects starting from this source
      _bind(event.source, index);
    }
  }

  void _setTargetProperty(Object propValue) {
    if (_valueTransform != null) {
      if (_transformArg != null) {
        propValue = _valueTransform(propValue, _transformArg);
      } else {
        propValue = _valueTransform(propValue, null);
      }
    }
    // set the value of the target.
    if (_targetProp is String) {
      String destPath = _targetProp;
      int pos = destPath.indexOf(".");
      int startPos = 0;
      IKeyValue target = _targetObject;
      PropertyDefinition propDef;
      while (pos != -1) {
        String pathElem = destPath.substring(startPos, pos);
        startPos = pos + 1;
        propDef = null;
        if (target is UxmlElement) {
          UxmlElement targetElm = target;
          propDef = targetElm.getDefinition().findProperty(pathElem);
        }
        if (propDef != null) {
          target = target.getProperty(propDef);
          if (target == null) {
            // Destination is not bound yet.
            // TODO(ferhat): should we listen to target being ready and
            // set property afterwards ?
            return; // Can't set property on nonexisting target.
          }
        } else {
          target = target.getProperty(pathElem);
          if (target == null) {
            // TODO(ferhat): see above.
            return;
          }
        }
        pos = destPath.indexOf(".", startPos);
      }
      if (target != null) {
        target.setProperty(destPath.substring(startPos), propValue);
      }
    } else {
      _targetObject.setProperty(_targetProp, _implicitConv(_targetProp,
          propValue));
    }
  }

  // Performs implicit type conversions.
  Object _implicitConv(PropertyDefinition prop, Object value) {
    if (prop != null) {
      switch(prop.dataType) {
        case PropertyType.STRING:
          if (!(value is String)) {
            return value == null ? "" : value.toString();
          }
          break;
        case PropertyType.INT:
          if (value is int) {
            return value;
          }
          if (value is String) {
            return int.parse(value);
          } else if (value is num) {
            num n = value;
            return n.toInt();
          } else if (value is bool) {
            bool b = value;
            return b ? 1 : 0;
          }
          break;
        case PropertyType.NUMBER:
          if (value is num) {
            return value;
          }
          if (value is String) {
            return double.parse(value);
          } else if (value is bool) {
            bool b = value;
            return b ? 1 : 0;
          }
          break;
        case PropertyType.BOOL:
          if (value is bool) {
            return value;
          }
          if (value is String) {
            String sVal = value;
            return !(sVal == "0" || sVal == "false");
          } else if (value is num) {
            num n = value;
            return n == 0 ? false : true;
          }
          break;
      }
    }
    return value;
  }

  /** BuiltIn value transform for boolean negate. */
  static Object negateBoolean(Object value, String transformArg) {
    bool boolVal = value;
    return !boolVal;
  }
}

typedef Object BindingTransform(Object value, String argument);
